### PR TITLE
chore: Auto-generates singular data source schemas in TPF resources

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Auto-generates singular data source schemas in TPF resources.

As an example it is used in data source schemas for employee_grant_access, push_log_based_export, search_deployment and advanced_cluster_tpf.

The data sources have auto-generated documentation, so PR checks would fail if these auto-generated data source schemas were not like the manual ones.

Link to any related issue(s): CLOUDP-283230

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals I have added appropriate changelog entries.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/encryptionatrest/resource.go [hunk 1]
```diff
@@ -52,202 +45,8 @@ type encryptionAtRestRS struct {
...
-type TFGcpKmsConfigModel struct {
-	ServiceAccountKey    types.String `tfsdk:"service_account_key"`
-	KeyVersionResourceID types.String `tfsdk:"key_version_resource_id"`
-	Enabled              types.Bool   `tfsdk:"enabled"`
-	Valid                types.Bool   `tfsdk:"valid"`
-}
-
 func (r *encryptionAtRestRS) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
-	resp.Schema = schema.Schema{
-		Attributes: map[string]schema.Attribute{
```
Comment (lantoli) [inline internal/service/encryptionatrest/resource.go:94]: move model and schema to resource_schema.go file

File: docs/resources/search_deployment.md [hunk 2]
```diff
@@ -59,7 +59,7 @@ output "mongodbatlas_search_deployment_id" {
 ### Required
 
 - `cluster_name` (String) Label that identifies the cluster to return the search nodes for.
-- `project_id` (String) Unique 24-hexadecimal character string that identifies the project.
+- `project_id` (String) Unique 24-hexadecimal digit string that identifies your project.
```
Comment (lantoli) [inline docs/resources/search_deployment.md:62]: there was a small mismatch between resource and data source schema documentation

File: internal/common/conversion/schema_description.go [hunk 3]
```diff
@@ -2,15 +2,104 @@ package conversion
...
+			continue
+		}
+		computed := true
+		required := false
+		if slices.Contains(requiredFields, name) {
+			computed = false
+			required = true
+		}
+		switch attrTyped := attr.(type) {
+		case schema.StringAttribute:
```
Comment (lantoli) [inline internal/common/conversion/schema_description.go]: i'll try to simplify this with reflection

**Review by carriecwk — APPROVED** [review]
lgtm

File: internal/service/mongodbemployeeaccessgrant/data_source.go [hunk 4]
```diff
@@ -24,8 +26,14 @@ type ds struct {
 }
 
 func (d *ds) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
-	resp.Schema = DataSourceSchema(ctx)
-	conversion.UpdateSchemaDescription(&resp.Schema)
+	// TODO: THIS WILL BE REMOVED BEFORE MERGING, check old data source schema and new auto-generated schema are the same
```
Comment (EspenAlbert) [inline internal/service/mongodbemployeeaccessgrant/data_source.go]: still TODO?

File: internal/service/mongodbemployeeaccessgrant/data_source_schema.go [hunk 5]
```diff
@@ -6,7 +6,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
 )
 
-func DataSourceSchema(ctx context.Context) schema.Schema {
+func DataSourceSchemaDelete(ctx context.Context) schema.Schema {
```
Comment (EspenAlbert) [inline internal/service/mongodbemployeeaccessgrant/data_source_schema.go]: Is this meant to be removed?

File: internal/service/pushbasedlogexport/data_source.go [hunk 6]
```diff
@@ -25,8 +27,14 @@ type pushBasedLogExportDS struct {
 }
 
 func (d *pushBasedLogExportDS) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
-	resp.Schema = DataSourceSchema(ctx)
-	conversion.UpdateSchemaDescription(&resp.Schema)
+	// TODO: THIS WILL BE REMOVED BEFORE MERGING, check old data source schema and new auto-generated schema are the same
```
Comment (EspenAlbert) [inline internal/service/pushbasedlogexport/data_source.go]: still todo?

**Review by EspenAlbert — APPROVED** [review]
Seems like a great solution!

File: internal/service/mongodbemployeeaccessgrant/data_source_schema.go (hunk 5 above)
Comment (lantoli) [reply internal/service/mongodbemployeeaccessgrant/data_source_schema.go]: this and the TODO will be removed, i'm seeing just before removing them if I can simplify convertAttrs with reflection

**Comment by lantoli** [comment]
@EspenAlbert conversions simplified here https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2799/commits/ff8985d7a0f690b08292dc13888db51e294660cf, and unneeded code removed here https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2799/commits/5c6bd8b743af5ad30c82bc288386e935327bf051

File: internal/common/conversion/schema_description.go (hunk 3 above)
Comment (lantoli) [reply internal/common/conversion/schema_description.go]: done

File: internal/common/conversion/schema_description.go [hunk 7]
```diff
@@ -2,15 +2,83 @@ package conversion
 
 import (
 	"reflect"
+	"slices"
```
Comment (AgustinBettati) [inline internal/common/conversion/schema_description.go]: nit: file is named `schema_description`, we could transition to a more general name like `schema_generation` now that this functionality is also present

File: internal/common/conversion/schema_description.go [hunk 8]
```diff
@@ -2,15 +2,83 @@ package conversion
...
 
 	dsschema "github.com/hashicorp/terraform-plugin-framework/datasource/schema"
 	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
 )
 
+func DataSourceSchemaFromResource(rs schema.Schema, requiredFields ...string) dsschema.Schema {
```
Comment (AgustinBettati) [inline internal/common/conversion/schema_description.go]: q: Are there any limitations worth calling out? I do see that for example project data source does have optional attributes which would not be supported.

**Review by AgustinBettati — APPROVED** [review]
LGTM, nice to see existing resources being simplified

File: internal/common/conversion/schema_description.go (hunk 7 above)
Comment (lantoli) [reply internal/common/conversion/schema_description.go]: changed, thx

File: internal/common/conversion/schema_description.go (hunk 8 above)
Comment (lantoli) [reply internal/common/conversion/schema_description.go]: at the moment blocks is not implemented (it panics so it's evident that needs to be implemented).  my idea is to add functionality as we need it.  Every time we use this func for existing data source schemas, we need to be sure that we're generating the same schema, as I did before this commit https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2799/commits/5c6bd8b743af5ad30c82bc288386e935327bf051, and implement any additional logic if needed.

Comment (lantoli) [reply internal/common/conversion/schema_description.go]: you're right in the case of optional params or with validators, we'll need to add some additional param to pass them

**Pull request merged**